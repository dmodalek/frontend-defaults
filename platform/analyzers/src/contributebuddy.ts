import { directoryExists, getDependencyInstallation, DependencyInstallation } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type ContributebuddyAnalyzerResult = {
    contributebuddy: boolean;
    contributebuddyConfigFolder?: string,
    contributebuddyInstallation?: DependencyInstallation
};

export const contributebuddyAnalyzer = async (cwd: string): Promise<ContributebuddyAnalyzerResult> => {
    const CONTRIBUTE_BUDDY_FOLDER = join(cwd, '.contributebuddy');
    const contributebuddyConfigFolder = await directoryExists(CONTRIBUTE_BUDDY_FOLDER);
    const contributebuddyInstallation = await getDependencyInstallation(cwd, 'contribute-buddy');

    if (contributebuddyConfigFolder || contributebuddyInstallation.declared) {
        return {
            contributebuddy: true,
            contributebuddyConfigFolder: CONTRIBUTE_BUDDY_FOLDER,
            contributebuddyInstallation

        }
    }

    return {
        contributebuddy: false,
    };
}