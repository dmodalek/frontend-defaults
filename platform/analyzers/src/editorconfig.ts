import { Analyzer, fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type EditorconfigAnalyzerResult = {
	editorConfig: boolean;
};

export class EditorconfigAnalyzer extends Analyzer<EditorconfigAnalyzerResult> {
	async analyze(): Promise<EditorconfigAnalyzerResult> {
		const eidtorConfigExists = await fileExists(join(this.context, '.editorconfig'));

		return {
			editorConfig: eidtorConfigExists,
		};
	}
}
