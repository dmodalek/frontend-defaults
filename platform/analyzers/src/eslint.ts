import { DependencyInstallation, fileExists, getDependencyInstallation } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type ESLintAnalyzerResult = {
	eslint: boolean;
	eslintIgnore?: boolean;
	eslintInstallation?: DependencyInstallation
};

export const eslintAnalyzer = async (cwd: string): Promise<ESLintAnalyzerResult> => {
	const doesESLintExist = await fileExists(join(cwd, '.eslintrc.js'));

	if (doesESLintExist) {
		const doesESLintIgnoreExist = await fileExists(join(cwd, '.eslintignore'));
		const eslintInstallation = await getDependencyInstallation(cwd, 'eslint');

		return {
			eslint: doesESLintExist,
			eslintIgnore: doesESLintIgnoreExist,
			eslintInstallation
		};
	}

	return {
		eslint: doesESLintExist,
	};
}