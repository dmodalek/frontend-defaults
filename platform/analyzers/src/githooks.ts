import {
	DependencyInstallation,
	fileExists,
	getDependencyInstallation,
	getPackageJSON,
	IPackage,
} from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type GitHooksLocation = 'rc' | 'package';

export type GitHooksAnalyzerResult = {
	gitHooks: boolean;
	gitHooksInstallation?: DependencyInstallation;
	gitHooksLocation?: GitHooksLocation;
};

type PackageWithHuskyConfig = IPackage & {
	husky: any;
};

export const gitHooksAnalyzer = async (cwd: string): Promise<GitHooksAnalyzerResult> => {
	const doesGitHooksConfigExist = await fileExists(join(cwd, '.huskyrc'));
	const gitHooksInstallation = await getDependencyInstallation(cwd, 'husky');
	const pkg: PackageWithHuskyConfig = (await getPackageJSON(cwd)) as any;

	if (doesGitHooksConfigExist || pkg.husky) {
		return {
			gitHooks: doesGitHooksConfigExist,
			gitHooksLocation: pkg.husky ? 'package' : 'rc',
			gitHooksInstallation,
		};
	}

	return {
		gitHooks: false,
	};
};
