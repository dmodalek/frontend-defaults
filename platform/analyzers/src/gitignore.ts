import { fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type GitIgnoreAnalyzerResult = {
	gitignore: boolean;
};

export const gitIgnoreAnalyzer = async (cwd: string): Promise<GitIgnoreAnalyzerResult> => {
	const doesGitignoreExist = await fileExists(join(cwd, '.gitignore'));

	return {
		gitignore: doesGitignoreExist,
	};
}