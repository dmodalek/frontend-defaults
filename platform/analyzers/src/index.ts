// export all related analyzer stuff for contributebuddy
export { contributebuddyAnalyzer, ContributebuddyAnalyzerResult } from './contributebuddy';

// export all related analyzer stuff for editorconfig
export { editorConfigAnalyzer, EditorconfigAnalyzerResult } from './editorconfig';

// export all related analyzer stuff for eslint
export { eslintAnalyzer, ESLintAnalyzerResult } from './eslint';

// export all related analyzer stuff for githooks
export { gitHooksAnalyzer, GitHooksAnalyzerResult } from './githooks';

// export all related analyzer stuff for gitignore
export { gitIgnoreAnalyzer, GitIgnoreAnalyzerResult } from './gitignore';

// export all related analyzer stuff for license
export { licenseAnalyzer, LicenseAnalyzerResult } from './license';

// export all related analyzer stuff for nitro
export { nitroAnalyzer, NitroAnalyzerResult } from './nitro';

// export all related analyzer stuff for node
export { nodeAnalyzer, NodeAnalyzerResult } from './node';

// export all related analyzer stuff for npmrc
export { npmrcAnalyzer, NPMRCAnalyzerResult } from './npmrc';

// export all related analyzer stuff for prettier
export { prettierAnalyzer, PrettierAnalyzerResult } from './prettier';

// export all related analyzer stuff for readme
export { readmeAnalyzer, ReadmeAnalyzerResult } from './readme';

// export all related analyzer stuff for repository
export { repositoryAnalyzer, RepositoryAnalyzerResult } from './repository';

// export all related analyzer stuff for stylelint
export { stylelintAnalyzer, StylelintAnalyzerResult } from './stylelint';

// export all related analyzer stuff for tslint
export { tslintAnalyzer, TSLintAnalyzerResult } from './tslint';

// export all related analyzer stuff for typescript
export { typescriptAnalyzer, TypeScriptAnalyzerResult } from './typescript';

// export all related analyzer stuff for webpack
export { webpackAnalyzer, WebpackAnalyzerResult } from './webpack';
