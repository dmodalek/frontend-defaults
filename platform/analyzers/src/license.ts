import { Analyzer, fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type LicenseAnalyzerResult = {
	license: boolean;
};

export class LicenseAnalyzer extends Analyzer<LicenseAnalyzerResult> {
	async analyze(): Promise<LicenseAnalyzerResult> {
		const licenseExists = await fileExists(join(this.context, 'LICENSE'));

		return {
			license: licenseExists,
		};
	}
}
