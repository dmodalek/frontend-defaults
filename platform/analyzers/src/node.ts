import { Analyzer, fileExists, getFileContents } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export enum NodeAnalyzerManagerType {
	'nvm' = 'nvm',
	'node-version' = 'node-version',
}

type NodeAnalyzerManagerInfoType = { [managerType in NodeAnalyzerManagerType]: string | false };

export type NodeAnalyzerResult = {
	nodeVersion: boolean;
	nodeVersionManagerInfo?: NodeAnalyzerManagerInfoType;
};

async function getLocalVersionFile(cwd: string): Promise<{ exist: boolean; managers: NodeAnalyzerManagerInfoType }> {
	const allManagers: NodeAnalyzerManagerInfoType = {
		[NodeAnalyzerManagerType['node-version']]: false as false,
		[NodeAnalyzerManagerType.nvm]: false as false,
	};
	const nodeVersionFileExists = await fileExists(join(cwd, '.node-version'));
	const nvmrcFileExists = await fileExists(join(cwd, '.nvmrc'));

	if (nodeVersionFileExists) {
		allManagers['node-version'] = await extractVersionFromFile(join(cwd, '.node-version'));
	}

	if (nvmrcFileExists) {
		allManagers.nvm = await extractVersionFromFile(join(cwd, '.nvmrc'));
	}

	return {
		exist: nvmrcFileExists || nodeVersionFileExists,
		managers: allManagers,
	};
}

async function extractVersionFromFile(filePath: string) {
	const contents = await getFileContents(filePath);
	return contents
		.replace(/\r?\n|\r/g, '')
		.replace('^', '')
		.replace('~', '')
		.replace('>', '')
		.replace('>=', '')
		.replace('<', '')
		.replace('<=', '');
}

export const nodeAnalyzer = async (cwd: string): Promise<NodeAnalyzerResult> => {
	const localVersionFiles = await getLocalVersionFile(cwd);

	if (localVersionFiles.exist) {
		return {
			nodeVersion: true,
			nodeVersionManagerInfo: localVersionFiles.managers,
		};
	}

	return {
		nodeVersion: false,
	};
};
