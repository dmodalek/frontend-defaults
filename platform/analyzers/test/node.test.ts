import { ProjectAnalyzer } from '@namics/frontend-defaults-platform-core';
import { getFixturePath } from './suite';
import { NodeAnalyzerResult, NodeAnalyzer } from '../src/node';

const FIXTURE_NODE_VERSION = getFixturePath('node-version-only-project');
const FIXTURE_NVM_VERSION = getFixturePath('nvm-only-project');
const FIXTURE_NONE = getFixturePath('default-project');
const FIXTURE_BOTH = getFixturePath('version-manager-project');

const getFixtureAnalyzer = async (context: string): Promise<ProjectAnalyzer<NodeAnalyzerResult>> => {
	return new ProjectAnalyzer<NodeAnalyzerResult>({
		context,
		analyzers: [NodeAnalyzer],
	}).boot();
};

describe('Analyzers', () => {
	describe('NodeAnalyzer', () => {
		it('should not crash the ProjectAnalyzer', async () => {
			expect(async () => await getFixtureAnalyzer(FIXTURE_NODE_VERSION)).not.toThrow();
			expect(async () => await getFixtureAnalyzer(FIXTURE_NVM_VERSION)).not.toThrow();
		});

		it('should analyze a project without node meta configuration', async () => {
			const analyzer = await getFixtureAnalyzer(FIXTURE_NONE);

			expect(analyzer.analytics.nodeVersion).toEqual(false);
			expect(analyzer.analytics).toMatchSnapshot();
		});

		it('should analyze a project with node-version correctly', async () => {
			const analyzer = await getFixtureAnalyzer(FIXTURE_NODE_VERSION);

			expect(analyzer.analytics.nodeVersion).toEqual(true);
			expect(analyzer.analytics.nodeVersionManagerInfo).toBeTruthy();
			expect(analyzer.analytics.nodeVersionManagerInfo!["node-version"]).toEqual('10.0.0');
			expect(analyzer.analytics).toMatchSnapshot();
		});

		it('should analyze a project with nvmrc correctly', async () => {
			const analyzer = await getFixtureAnalyzer(FIXTURE_NVM_VERSION);

			expect(analyzer.analytics.nodeVersion).toEqual(true);
			expect(analyzer.analytics.nodeVersionManagerInfo).toBeTruthy();
			expect(analyzer.analytics.nodeVersionManagerInfo!.nvm).toEqual('10.0.0');
			expect(analyzer.analytics).toMatchSnapshot();
		});
	});
});
