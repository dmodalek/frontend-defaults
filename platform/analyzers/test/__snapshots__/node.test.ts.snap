// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analyzers NodeAnalyzer should analyze a project with node-version correctly 1`] = `
Object {
  "nodeVersion": true,
  "nodeVersionManagerInfo": Object {
    "node-version": "10.0.0",
    "nvm": false,
  },
}
`;

exports[`Analyzers NodeAnalyzer should analyze a project with nvmrc correctly 1`] = `
Object {
  "nodeVersion": true,
  "nodeVersionManagerInfo": Object {
    "node-version": false,
    "nvm": "10.0.0",
  },
}
`;

exports[`Analyzers NodeAnalyzer should analyze a project without node meta configuration 1`] = `
Object {
  "nodeVersion": false,
}
`;
